# Exploratory Data Analysis


## Importing the libraries

In [1]:
import pandas as pd

## Data Cleaning - Duplicate Data
### Sample Sales Data

This dataset is randomly generated.

In [15]:
sales_df = pd.read_csv('./data/sales/sales.csv')

In [16]:
# Get the first 5 records
sales_df.head()

Order ID    Product  Quantity  Price Customer Name       Address State  \
0      1001  Product A         2  10.99    Customer 1   123 Main St    OH   
1      1002  Product B         1    NaN    Customer 2    456 Elm St    OH   
2      1003  Product A         3  10.99    Customer 3    789 Oak St    KY   
3      1004  Product C         1  25.99    Customer 4  1011 Pine St    OH   
4      1005  Product B         2  15.99    Customer 1   123 Main St    OH   

         Date  
0  2022-01-01  
1  2022-01-02  
2  2022-01-03  
3  2022-01-04  
4  2022-01-05

In [17]:
# Get the last 5 records
sales_df.tail()

Order ID    Product  Quantity  Price Customer Name      Address State  \
17      1017  Product A         2  10.99           NaN   789 Oak St    KY   
18      1018  Product C         1  25.99    Customer 1  123 Main St    OH   
19      1019  Product B         3  15.99    Customer 2   456 Elm St    OH   
20      1020  Product F         2  10.99    Customer 1  123 Main St    OH   
21      1021  Product F         2  10.99    Customer 1  123 Main St    OH   

          Date  
17  2022-01-17  
18  2022-01-18  
19         NaN  
20  2022-01-19  
21  2022-01-19

In [19]:
# Get 5 random records
sales_df.sample(5)

Order ID    Product  Quantity  Price Customer Name       Address State  \
21      1021  Product F         2  10.99    Customer 1   123 Main St    OH   
5       1006  Product A         1  10.99    Customer 2    456 Elm St    OH   
16      1016  Product E         3  45.99    Customer 4  1011 Pine St    OH   
9       1009  Product E         2  45.99    Customer 1   123 Main St    OH   
18      1018  Product C         1  25.99    Customer 1   123 Main St    OH   

          Date  
21  2022-01-19  
5   2022-01-06  
16  2022-01-16  
9   2022-01-09  
18  2022-01-18

In [13]:
# how many records are there?
sales_df.shape

(22, 8)

<details>
<summary><strong>Hints on where the duplicate data are</strong></summary>
  
  * 1007 is there twice (same order ID)
  * 1020, 1021, different IDs but same everything else (merge?)
    * 'Product', 'Quantity', 'Date','Customer Name'
</details>

In [21]:
# Exact duplicates
sales_df.duplicated().sum()

1

In [23]:
# Duplicates based on a subset of columns (sum)
sales_df.duplicated(
    subset=['Product', 'Quantity', 'Customer Name', 'Date']
).sum()


2

In [ ]:
# Duplicates based on a subset of columns (all)
sales_df.duplicated(
    subset=['Product', 'Quantity', 'Customer Name', 'Date']
)

In [25]:
# Control Which one is the duplicate
sales_df.duplicated(
    subset=['Product', 'Quantity', 'Customer Name', 'Date'],
    keep='last'
)

0     False
1     False
2     False
3     False
4     False
5     False
6      True
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20     True
21    False
dtype: bool

In [34]:
# See the actual records that are marked as duplicate
sales_df[sales_df.duplicated(
    subset=['Product']
)]

Order ID    Product  Quantity  Price Customer Name       Address State  \
2       1003  Product A         3  10.99    Customer 3    789 Oak St    KY   
4       1005  Product B         2  15.99    Customer 1   123 Main St    OH   
5       1006  Product A         1  10.99    Customer 2    456 Elm St    OH   
7       1007  Product D         2  35.99    Customer 3    789 Oak St    OH   
8       1008  Product B         1  15.99    Customer 4  1011 Pine St    OH   
10      1010  Product A         3  10.99    Customer 2    456 Elm St    OH   
11      1011  Product B         1  15.99    Customer 3    789 Oak St    KY   
12      1012  Product C         2  25.99           NaN  1011 Pine St    OH   
13      1013  Product A         1  10.99    Customer 1   123 Main St    OH   
14      1014  Product D         2    NaN    Customer 2    456 Elm St    OH   
15      1015  Product B         1  15.99    Customer 3           NaN    OH   
16      1016  Product E         3  45.99    Customer 4  1011 Pine St    OH   
17      1017  Product A         2  10.99           NaN    789 Oak St    KY   
18      1018  Product C         1  25.99    Customer 1   123 Main St    OH   
19      1019  Product B         3  15.99    Customer 2    456 Elm St    OH   
21      1021  Product F         2  10.99    Customer 1   123 Main St    OH   

          Date  
2   2022-01-03  
4   2022-01-05  
5   2022-01-06  
7   2022-01-07  
8   2022-01-08  
10  2022-01-10  
11  2022-01-11  
12  2022-01-12  
13  2022-01-13  
14  2022-01-14  
15  2022-01-15  
16  2022-01-16  
17  2022-01-17  
18  2022-01-18  
19         NaN  
21  2022-01-19

In [36]:
sales_df[
    sales_df['Product'] == 'Product A'
]

Order ID    Product  Quantity  Price Customer Name      Address State  \
0       1001  Product A         2  10.99    Customer 1  123 Main St    OH   
2       1003  Product A         3  10.99    Customer 3   789 Oak St    KY   
5       1006  Product A         1  10.99    Customer 2   456 Elm St    OH   
10      1010  Product A         3  10.99    Customer 2   456 Elm St    OH   
13      1013  Product A         1  10.99    Customer 1  123 Main St    OH   
17      1017  Product A         2  10.99           NaN   789 Oak St    KY   

          Date  
0   2022-01-01  
2   2022-01-03  
5   2022-01-06  
10  2022-01-10  
13  2022-01-13  
17  2022-01-17

In [37]:
sales_df[
    sales_df['Price'] > 10.99
]

Order ID    Product  Quantity  Price Customer Name       Address State  \
3       1004  Product C         1  25.99    Customer 4  1011 Pine St    OH   
4       1005  Product B         2  15.99    Customer 1   123 Main St    OH   
6       1007  Product D         2  35.99    Customer 3    789 Oak St    OH   
7       1007  Product D         2  35.99    Customer 3    789 Oak St    OH   
8       1008  Product B         1  15.99    Customer 4  1011 Pine St    OH   
9       1009  Product E         2  45.99    Customer 1   123 Main St    OH   
11      1011  Product B         1  15.99    Customer 3    789 Oak St    KY   
12      1012  Product C         2  25.99           NaN  1011 Pine St    OH   
15      1015  Product B         1  15.99    Customer 3           NaN    OH   
16      1016  Product E         3  45.99    Customer 4  1011 Pine St    OH   
18      1018  Product C         1  25.99    Customer 1   123 Main St    OH   
19      1019  Product B         3  15.99    Customer 2    456 Elm St    OH   

          Date  
3   2022-01-04  
4   2022-01-05  
6   2022-01-07  
7   2022-01-07  
8   2022-01-08  
9   2022-01-09  
11  2022-01-11  
12  2022-01-12  
15  2022-01-15  
16  2022-01-16  
18  2022-01-18  
19         NaN

In [26]:
sales_df.shape

(22, 8)

#### How to deal? Drop

In [27]:
# Drop the exact duplicates
sales_df.drop_duplicates()

Order ID    Product  Quantity  Price Customer Name       Address State  \
0       1001  Product A         2  10.99    Customer 1   123 Main St    OH   
1       1002  Product B         1    NaN    Customer 2    456 Elm St    OH   
2       1003  Product A         3  10.99    Customer 3    789 Oak St    KY   
3       1004  Product C         1  25.99    Customer 4  1011 Pine St    OH   
4       1005  Product B         2  15.99    Customer 1   123 Main St    OH   
5       1006  Product A         1  10.99    Customer 2    456 Elm St    OH   
6       1007  Product D         2  35.99    Customer 3    789 Oak St    OH   
8       1008  Product B         1  15.99    Customer 4  1011 Pine St    OH   
9       1009  Product E         2  45.99    Customer 1   123 Main St    OH   
10      1010  Product A         3  10.99    Customer 2    456 Elm St    OH   
11      1011  Product B         1  15.99    Customer 3    789 Oak St    KY   
12      1012  Product C         2  25.99           NaN  1011 Pine St    OH   
13      1013  Product A         1  10.99    Customer 1   123 Main St    OH   
14      1014  Product D         2    NaN    Customer 2    456 Elm St    OH   
15      1015  Product B         1  15.99    Customer 3           NaN    OH   
16      1016  Product E         3  45.99    Customer 4  1011 Pine St    OH   
17      1017  Product A         2  10.99           NaN    789 Oak St    KY   
18      1018  Product C         1  25.99    Customer 1   123 Main St    OH   
19      1019  Product B         3  15.99    Customer 2    456 Elm St    OH   
20      1020  Product F         2  10.99    Customer 1   123 Main St    OH   
21      1021  Product F         2  10.99    Customer 1   123 Main St    OH   

          Date  
0   2022-01-01  
1   2022-01-02  
2   2022-01-03  
3   2022-01-04  
4   2022-01-05  
5   2022-01-06  
6   2022-01-07  
8   2022-01-08  
9   2022-01-09  
10  2022-01-10  
11  2022-01-11  
12  2022-01-12  
13  2022-01-13  
14  2022-01-14  
15  2022-01-15  
16  2022-01-16  
17  2022-01-17  
18  2022-01-18  
19         NaN  
20  2022-01-19  
21  2022-01-19

In [28]:
# How many records are there now?
sales_df.shape

(22, 8)

In [29]:
sales_df1 = sales_df.drop_duplicates()
sales_df1.shape

(21, 8)

In [ ]:
# sales_df.drop_duplicates(inplace=True)
# sales_df1.shape

In [38]:
# Drop the semi-duplicates
sales_df2 = sales_df.drop_duplicates(
    subset=['Product', 'Quantity', 'Customer Name', 'Date'],
    keep='last'
)

In [39]:
# How many records are there now?
sales_df2.shape

(20, 8)

> Talk about the inplace parameter

#### How to deal? Merge

In [40]:
sales_df

Order ID    Product  Quantity  Price Customer Name       Address State  \
0       1001  Product A         2  10.99    Customer 1   123 Main St    OH   
1       1002  Product B         1    NaN    Customer 2    456 Elm St    OH   
2       1003  Product A         3  10.99    Customer 3    789 Oak St    KY   
3       1004  Product C         1  25.99    Customer 4  1011 Pine St    OH   
4       1005  Product B         2  15.99    Customer 1   123 Main St    OH   
5       1006  Product A         1  10.99    Customer 2    456 Elm St    OH   
6       1007  Product D         2  35.99    Customer 3    789 Oak St    OH   
7       1007  Product D         2  35.99    Customer 3    789 Oak St    OH   
8       1008  Product B         1  15.99    Customer 4  1011 Pine St    OH   
9       1009  Product E         2  45.99    Customer 1   123 Main St    OH   
10      1010  Product A         3  10.99    Customer 2    456 Elm St    OH   
11      1011  Product B         1  15.99    Customer 3    789 Oak St    KY   
12      1012  Product C         2  25.99           NaN  1011 Pine St    OH   
13      1013  Product A         1  10.99    Customer 1   123 Main St    OH   
14      1014  Product D         2    NaN    Customer 2    456 Elm St    OH   
15      1015  Product B         1  15.99    Customer 3           NaN    OH   
16      1016  Product E         3  45.99    Customer 4  1011 Pine St    OH   
17      1017  Product A         2  10.99           NaN    789 Oak St    KY   
18      1018  Product C         1  25.99    Customer 1   123 Main St    OH   
19      1019  Product B         3  15.99    Customer 2    456 Elm St    OH   
20      1020  Product F         2  10.99    Customer 1   123 Main St    OH   
21      1021  Product F         2  10.99    Customer 1   123 Main St    OH   

          Date  
0   2022-01-01  
1   2022-01-02  
2   2022-01-03  
3   2022-01-04  
4   2022-01-05  
5   2022-01-06  
6   2022-01-07  
7   2022-01-07  
8   2022-01-08  
9   2022-01-09  
10  2022-01-10  
11  2022-01-11  
12  2022-01-12  
13  2022-01-13  
14  2022-01-14  
15  2022-01-15  
16  2022-01-16  
17  2022-01-17  
18  2022-01-18  
19         NaN  
20  2022-01-19  
21  2022-01-19

In [47]:
# Merge the semi-duplicates
sales_df_grouped = sales_df1.groupby(
    ['Product', 'Quantity', 'Date', 'Customer Name'],
    dropna=False
).agg(
    {
        'Order ID': 'min',
        'Price': 'sum',
        'Address': 'first',
        'State': 'first'
    }
).sort_values('Order ID').reset_index().reindex(columns=sales_df.columns)
sales_df_grouped

Order ID    Product  Quantity  Price Customer Name       Address State  \
0       1001  Product A         2  10.99    Customer 1   123 Main St    OH   
1       1002  Product B         1   0.00    Customer 2    456 Elm St    OH   
2       1003  Product A         3  10.99    Customer 3    789 Oak St    KY   
3       1004  Product C         1  25.99    Customer 4  1011 Pine St    OH   
4       1005  Product B         2  15.99    Customer 1   123 Main St    OH   
5       1006  Product A         1  10.99    Customer 2    456 Elm St    OH   
6       1007  Product D         2  35.99    Customer 3    789 Oak St    OH   
7       1008  Product B         1  15.99    Customer 4  1011 Pine St    OH   
8       1009  Product E         2  45.99    Customer 1   123 Main St    OH   
9       1010  Product A         3  10.99    Customer 2    456 Elm St    OH   
10      1011  Product B         1  15.99    Customer 3    789 Oak St    KY   
11      1012  Product C         2  25.99           NaN  1011 Pine St    OH   
12      1013  Product A         1  10.99    Customer 1   123 Main St    OH   
13      1014  Product D         2   0.00    Customer 2    456 Elm St    OH   
14      1015  Product B         1  15.99    Customer 3          None    OH   
15      1016  Product E         3  45.99    Customer 4  1011 Pine St    OH   
16      1017  Product A         2  10.99           NaN    789 Oak St    KY   
17      1018  Product C         1  25.99    Customer 1   123 Main St    OH   
18      1019  Product B         3  15.99    Customer 2    456 Elm St    OH   
19      1020  Product F         2  21.98    Customer 1   123 Main St    OH   

          Date  
0   2022-01-01  
1   2022-01-02  
2   2022-01-03  
3   2022-01-04  
4   2022-01-05  
5   2022-01-06  
6   2022-01-07  
7   2022-01-08  
8   2022-01-09  
9   2022-01-10  
10  2022-01-11  
11  2022-01-12  
12  2022-01-13  
13  2022-01-14  
14  2022-01-15  
15  2022-01-16  
16  2022-01-17  
17  2022-01-18  
18         NaN  
19  2022-01-19

Probably we should use a mix of both. For exact, drop, for semi, merge.

## Data Cleaning - Missing Data
### Sample Sales Data

[📜 Pandas `isna()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isna.html#pandas.DataFrame.isna)

In [48]:
sales_df_grouped

Order ID    Product  Quantity  Price Customer Name       Address State  \
0       1001  Product A         2  10.99    Customer 1   123 Main St    OH   
1       1002  Product B         1   0.00    Customer 2    456 Elm St    OH   
2       1003  Product A         3  10.99    Customer 3    789 Oak St    KY   
3       1004  Product C         1  25.99    Customer 4  1011 Pine St    OH   
4       1005  Product B         2  15.99    Customer 1   123 Main St    OH   
5       1006  Product A         1  10.99    Customer 2    456 Elm St    OH   
6       1007  Product D         2  35.99    Customer 3    789 Oak St    OH   
7       1008  Product B         1  15.99    Customer 4  1011 Pine St    OH   
8       1009  Product E         2  45.99    Customer 1   123 Main St    OH   
9       1010  Product A         3  10.99    Customer 2    456 Elm St    OH   
10      1011  Product B         1  15.99    Customer 3    789 Oak St    KY   
11      1012  Product C         2  25.99           NaN  1011 Pine St    OH   
12      1013  Product A         1  10.99    Customer 1   123 Main St    OH   
13      1014  Product D         2   0.00    Customer 2    456 Elm St    OH   
14      1015  Product B         1  15.99    Customer 3          None    OH   
15      1016  Product E         3  45.99    Customer 4  1011 Pine St    OH   
16      1017  Product A         2  10.99           NaN    789 Oak St    KY   
17      1018  Product C         1  25.99    Customer 1   123 Main St    OH   
18      1019  Product B         3  15.99    Customer 2    456 Elm St    OH   
19      1020  Product F         2  21.98    Customer 1   123 Main St    OH   

          Date  
0   2022-01-01  
1   2022-01-02  
2   2022-01-03  
3   2022-01-04  
4   2022-01-05  
5   2022-01-06  
6   2022-01-07  
7   2022-01-08  
8   2022-01-09  
9   2022-01-10  
10  2022-01-11  
11  2022-01-12  
12  2022-01-13  
13  2022-01-14  
14  2022-01-15  
15  2022-01-16  
16  2022-01-17  
17  2022-01-18  
18         NaN  
19  2022-01-19

In [49]:
# Find missing data (is null/is na)\
sales_df_grouped.isna()


Order ID  Product  Quantity  Price  Customer Name  Address  State   Date
0      False    False     False  False          False    False  False  False
1      False    False     False  False          False    False  False  False
2      False    False     False  False          False    False  False  False
3      False    False     False  False          False    False  False  False
4      False    False     False  False          False    False  False  False
5      False    False     False  False          False    False  False  False
6      False    False     False  False          False    False  False  False
7      False    False     False  False          False    False  False  False
8      False    False     False  False          False    False  False  False
9      False    False     False  False          False    False  False  False
10     False    False     False  False          False    False  False  False
11     False    False     False  False           True    False  False  False
12     False    False     False  False          False    False  False  False
13     False    False     False  False          False    False  False  False
14     False    False     False  False          False     True  False  False
15     False    False     False  False          False    False  False  False
16     False    False     False  False           True    False  False  False
17     False    False     False  False          False    False  False  False
18     False    False     False  False          False    False  False   True
19     False    False     False  False          False    False  False  False

In [51]:
# Again with Sum
sales_df.isna().sum()


Order ID         0
Product          0
Quantity         0
Price            2
Customer Name    2
Address          1
State            0
Date             1
dtype: int64

#### How to deal? Drop

In [52]:
sales_df3= sales_df.dropna()

In [56]:
sales_df3.isna().sum()
sales_df3.shape

(16, 8)

#### How to deal? Substitute with correct values

In [ ]:
sales_df

In [ ]:
# Get the rows with missing data


In [ ]:
# Fill NA with a map

In [ ]:
# Check again
sales_df[sales_df.isna().any(axis=1)]

#### How to deal? Substitute with an approximation

#### Side: Lambda Functions
Functions that have one one expression.

In [ ]:
## Origin
sales_df['State'] = sales_df['State'].map({'OH': 'Ohio', 'KY': 'Kentucky'})

#### Data Types

<details>
<summary><strong>Expand for a visual summary of the Data Types</strong></summary>
  
  ![Data types](https://miro.medium.com/max/1400/1*kySPZcf83qLOuaqB1vJxlg.jpeg)
</details>